microprofile/connectors/kafka/src/main/java/io/helidon/microprofile/connectors/kafka/KafkaMessage.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (c) 2020 Oracle and/or its affiliates. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.helidon.microprofile.connectors.kafka;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.eclipse.microprofile.reactive.messaging.Message;
+
+/**
+ * Kafka specific MP messaging message.
+ *
+ * @param <K> kafka record key type
+ * @param <V> kafka record value type
+ */
+class KafkaMessage<K, V> implements Message<ConsumerRecord<K, V>> {
+
+    private static final Logger LOGGER = Logger.getLogger(KafkaMessage.class.getName());
+    private final ConsumerRecord<K, V> consumerRecord;
+    private final AtomicBoolean ack = new AtomicBoolean(false);
+    private final CountDownLatch waitForCommit;
+    private final long millisWaitingTimeout;
+    private final AtomicReference<Exception> ackException = new AtomicReference<>();
+
+    /**
+     * Kafka specific MP messaging message.
+     *
+     * @param consumerRecord {@link org.apache.kafka.clients.consumer.ConsumerRecord}
+     * @param autoCommit when false it will ack will wait till it is really commited in Kafka,
+     *        otherwise there is no waiting time because it was committed already.
+     * @param millisWaitingTimeout this is the time in milliseconds that the ack will be waiting
+     *        the commit in Kafka. Applies only if autoCommit is false.
+     */
+    KafkaMessage(ConsumerRecord<K, V> consumerRecord, boolean autoCommit, long millisWaitingTimeout) {
+        this.consumerRecord = consumerRecord;
+        this.waitForCommit = new CountDownLatch(autoCommit ? 0 : 1);
+        this.millisWaitingTimeout = millisWaitingTimeout;
+    }
+
+    @Override
+    public ConsumerRecord<K, V> getPayload() {
+        return consumerRecord;
+    }
+
+    @Override
+    public CompletionStage<Void> ack() {
+        ack.set(true);
+        return CompletableFuture.runAsync(() -> {
```

Comment:
ack is going to be called frequently, can we avoid spinning up ForkJoinPool and reuse emit loop we already have? I know it introduces another queue